node_modules/col-canvas/lib/poller.js

```diff
@@ -553,9 +553,32 @@ var pollAssignments = function(ctx, users, opts, callback) {
         return callback(err);
       }
 
-      // Get the submissions for the assignment
+      // Handle submissions for each assignment
       var assignmentIterator = handleAssignment.bind(null, ctx, users, categories, opts);
-      async.eachSeries(assignments, assignmentIterator, callback);
+      async.eachSeries(assignments, assignmentIterator, function() {
+
+        // Re-fetch categories
+        CategoriesAPI.getCategories(ctx, true, true, function(err, categories) {
+          if (err) {
+            return callback(err);
+          }
+
+          // Remove any empty categories no longer corresponding to an active assignment. (The usual paranoid-mode delete isn't sufficient
+          // because that will prevent the category from being recreated if the assignment is later restored.)
+          var categoriesToDelete = _.filter(categories, function(c) {
+            return (c.canvas_assignment_id && (c.dataValues.asset_count === 0) && !_.find(assignments, {'id': c.canvas_assignment_id}))
+          });
+
+          async.eachSeries(categoriesToDelete, function(category, done) {
+            CategoriesAPI.deleteCategory(ctx, category.id, {'force': true}, function(err) {
+              if (err) {
+                log.error({'category': category.id, 'err': err}, 'Failed to delete a category');
+              }
+              return done();
+            });
+          }, callback);
```

Comment:
@paulkerschen  I'm very likely unclear...  In the previous impl, `callback` was invoked once: at finish of `eachSeries`.  In this new impl, we invoke once per iteration of outer `eachSeries`?

Reply:
@johncrossman This is why async is awful. The outer `eachSeries` invokes `assignmentIterator` once per assignment, and the completion function (lines 558-579) afterward. The inner `eachSeries` invokes the iteration function (lines 573-578) once per category, and `callback` afterward.